softioc/autosave.py

```diff
@@ -0,0 +1,131 @@
+        self.get_autosave_pvs()
+        self._state = {}
+        self._last_saved_state = {}
+        self._started = False
+        if self.enabled:
+            self.load()  # load at startup if enabled
+
+    def get_autosave_pvs(self):
+        self._pvs = {name: pv for name, pv in LookupRecordList() if pv.autosave}
+
+    def _change_directory(self, directory: str):
+        dir_path = Path(directory)
+        if dir_path.is_dir():
+            self.directory = dir_path
+        else:
+            raise ValueError(f"{directory} is not a valid autosave directory")
+
+    def backup_sav_file(self):
+        sav_path = self._get_current_sav_path()
+        if sav_path.is_file():
+            shutil.copy2(sav_path, self._get_timestamped_backup_sav_path())
+
+    def add_pv(self, pv):
+        pv.autosave = True
+        self._pvs[pv.name] = pv
+
+    def remove_pv(self, pv):
+        pv.autosave = False
+        self._pvs.pop(pv.name, None)
+
+    def _get_timestamped_backup_sav_path(self):
+        sav_path = self._get_current_sav_path()
+        return sav_path.parent / (
+            sav_path.name + self._last_saved_time.strftime("_%y%m%d-%H%M%S")
+        )
+
+    def _get_backup_save_path(self):
+        return self.directory / f"{self.device_name}.{SAVB_SUFFIX}"
+
+    def _get_current_sav_path(self):
+        return self.directory / f"{self.device_name}.{SAV_SUFFIX}"
+
+    def _save(self, state):
+        try:
+            for path in [self._get_current_sav_path(), self._get_backup_save_path()]:
+                with open(path, "w") as f:
+                    json.dump(state, f, indent=4)
+            self._last_saved_state = state.copy()  # do we need to copy?
+            self._last_saved_time = datetime.now()
+        except Exception as e:
+            print(f"Could not save state to file: {e}")
+
+    def save(self):
+        if not self.enabled:
+            print("Not saving to file as autosave adapter disabled")
+            return
+        state = {name: pv.get() for name, pv in self._pvs.items()}
+        if state != self._last_saved_state:
+            self._save(state)
+
+    def load(self, path = None):
+        if not self.enabled:
+            print("Not loading from file as autosave adapter disabled")
+            return
+        sav_path = path or self._get_current_sav_path()
+        if not sav_path or not sav_path.is_file():
+            print(f"Could not load autosave values from file {sav_path}")
+            return
+        with open(sav_path, "r") as f:
+            state = json.load(f)
+        for name, value in state.items():
+            pv = self._pvs.get(name, None)
+            if not pv:
+                print(f"{name} is not a valid autosaved PV")
+                continue
+            pv.set(value)
+
+    def loop(self):
+        if not self._pvs:
+            return  # end thread if no PVs to save
+        while True:
+            time.sleep(self.save_period)
+            self.save()
```

Comment:
E.g. if `self._stop_event = threading.Event()` in `__init__`
```suggestion
    def stop(self):
        self._stop_event.set()
        
    def loop(self):
        if not self._pvs:
            return  # end thread if no PVs to save
        while True:
            try:
                self._stop_event.wait(timeout=self.save_period)
            except TimeoutError:
                # No stop requested, we should save and continue
                self.save()
            else:
                # Stop requested
                return
```